### ISSUE TITLE
[verilator] Verilator does not support an interface for "host reset".

### ISSUE THREAD

**Issue by mcy** [lowRISC/opentitan#1028]
Right now, there's no way for the host to hold a running Verilator simulation in reset (for example, spiflash needs to hold the device in reset while it flashes a binary).

Unfortunately, this is a bit beyond what I'm capable of doing (since we'll probably have to introduce a new DPI for this).

**Comment by mcy** [comment]
@eunchan @moidx for visibility.

**Comment by imphil** [comment]
Not the most convenient option, but you can connect over JTAG with OpenOCD and hold the core in reset.

If spiflash needs to reset the device while flashing, that probably should be a functionality of the SPI interface? Or in other words: (why) is this only needed on Verilator?

**Comment by moidx** [comment]
In the FPGA we reset the device via FTDI. We just need to toggle reset to make sure `boot_rom` sees the bootstrap request.

**Comment by imphil** [comment]
Oh, so you're essentially looking for a Verilator version of the BOOTSTRAP/boot_strap pin? Since this is a normal GPIO pin, can't we just map that to the GPIO DPI module that we have in Verilator?

**Comment by mcy** [comment]
Not quite; I have a WIP change that makes spiflash drive GPI in Verilator mode. This is current behavior:
- I boot the Verilated system.
- Some unspecified state is observed in the BOOTSTRAP pin, so the boot rom figures no bootstrap is necessary.
- The simulation jumps into flash, and either:
  - Executes as normal, because there's a program there.
  - Freaks out at whatever garbage is in flash, jumps to an exception handler and loops forever.

In addition to driving the BOOTSTRAP pin to the appropriate voltage, we need a way to hold the system in reset so that it will execute from the beginning after we've started sending spiflash frames.

**Comment by imphil** [comment]
I'm still trying to figure out what makes Verilator different from the FPGA here. Is it the missing pullup/pulldown on the bootstrap pin?

**Comment by mcy** [comment]
@imphil Right now we have this hack for Verilator: https://github.com/lowRISC/opentitan/blob/master/sw/device/boot_rom/bootstrap.c#L20. My WIP change deletes this and has the Verilated device read the GPIO pins instead.

The problem seems to be that, regardless of whether a flash binary is passed to the simulation, we get the same pin voltages as in the gpio_in register, so we really have two options:
- Give spiflash-verilator a way to reset the device after pulling the bootstrap pin up.
- Have the simulation itself pull the bootstrap pin up when no flash is preloaded.

I think we want the first option, if only because that way we can spiflash a running simulator with the same semantics as spiflashing a running FPGA.

**Comment by moidx** [comment]
@imphil, in the FPGA we reset Ibex via FTDI while toggling the bootstrap pin. This ensures that boot_rom will see the bootstrap pin high out of reset. In Verilator we'd like to do the same, reset Ibex via a dpi wrapped reset pin to emulate the same behavior as with the FPGA.

**Comment by tjaychen** [comment]
@mcy are you actually trying to support spiflash in verilator? Because if you provide an image via flashinit, i dont think you actually need to bother.

but yeah... in FPGA, we can directly manipulate the IO_RSTN line, there's no equivalent in verilator at the moment.  When the simulation is instantiated, there IS a top level rst_ni signal, but it's not clear to me if you can directly manipulate that from somewhere else.

**Comment by rswarbrick** [comment]
@mcy / @moidx: Is this issue still live? I see that we still have an "is flash blank?" check to support bootstrap in the test rom. Is that the hack you were talking about? Does it actually matter?

**Comment by msfschaffner** [comment]
@alphan @arunthomas is this issue still relevant or can we close it?

**Comment by alphan** [comment]
Depends, do we want to support the bootstrap flow in verilator? I don't have a use case for this atm.
